### PR TITLE
feat :  Implemented Logic for Filtering Requests by State and Username

### PR DESCRIPTION
**Pull request by AnujChhikara**
Date: 27 Feb, 2025

Developer Name: @AnujChhikara 

---

## Issue Ticket Number

- #864 
## Description
This PR adds the logic to filter requests based on state and username. The UI for filtering was already in place, but the filtering functionality was not implemented. With this update:

- Users can now filter requests by state (e.g., pending, approved, rejected).
- Users can also filter requests by username, allowing them to quickly find relevant requests.
- Debouncing implemented: When a user types a username, we wait 300ms before sending the request to minimize API calls and improve performance.

### Documentation Updated?

- [ ] Yes
- [x] No

### Under Feature Flag

- [x] Yes
- [ ] No

### Database Changes

- [ ] Yes
- [x] No

### Breaking Changes

- [ ] Yes
- [x] No

### Development Tested?

- [x] Yes
- [ ] No

## Screenshots
<details>
<summary>Screenshot 1</summary>

 - Filter by State

https://github.com/user-attachments/assets/c5153b17-e937-450e-b147-ee3d30af17a6

 - Filter by Username

https://github.com/user-attachments/assets/37c687de-12d9-490c-8056-7d7d0f15d0e1

 - Approving and rejecting Requests

https://github.com/user-attachments/assets/486b161e-ceae-4b3d-913a-604449ab1d75

</details>

## Test Coverage
<details>
<summary>Screenshot 1</summary>

image

</details>

## Additional Notes
- This is the last PR for this task 

## Summary by CodeRabbit

- **New Features**
  - Introduced a new user suggestion feature based on the search term "randhir".
  - Improved request filtering to display only approved requests when the filter is applied.
  - Enhanced assignee search with a debounced input, providing timely and relevant user suggestions.

- **Style**
  - Refined the appearance of suggestion elements with adjusted spacing and font size for better readability.

### CODE REVIEW

File: requests/util.js [hunk 1]
```diff
@@ -43,6 +43,9 @@ function getQueryParamsString(requestType, query) {
   ) {
     queryParam += `&state=${query.state}`;
   }
+  if (query.requestedBy !== undefined && query.requestedBy !== null) {
+    queryParam += `&requestedBy=${query.requestedBy}`;
+  }
```
Comment (MayankBansal12) [inline requests/util.js:48]:
```suggestion
  if (query?.requestedBy != null) {
    queryParam += `&requestedBy=${query.requestedBy}`;
  }
```

File: mock-data/requests/index.js [hunk 2]
```diff
@@ -20,20 +20,18 @@ const pendingRequest = {
...
+    createdAt: 1711439903761,
+    requestedBy: 'V4rqL1aDecNGoa1IxiCu',
+    from: 1712275200000,
+    until: 1712448000000,
+    type: 'OOO',
+    message: 'request message',
+    lastModifiedBy: 'V4rqL1aDecNGoa1IxiCu',
+    state: 'APPROVED',
+    updatedAt: 1711482912686,
+  },
```
Comment (MayankBansal12) [inline mock-data/requests/index.js:34]: will it not return an array? if not, then why have next and prev cursor?

Comment (AnujChhikara) [reply mock-data/requests/index.js:34]: This mock data is used to fetch a specific request that is being approved or rejected. The API ensures that when a superuser approves or rejects a request, we fetch only that particular request and update it accordingly. This is the [API](https://api.realdevsquad.com/requests?dev=true&id=QjLMT6j1jsAdlr8WWa6B) example.

Comment (MayankBansal12) [reply mock-data/requests/index.js:34]: Understood. The mock data has prev and next cursor along with data, i guess that can be removed now!

**Review by MayankBansal12 — APPROVED** [review]

**Review by RishiChaubey31 — DISMISSED** [review]

File: requests/script.js [hunk 3]
```diff
@@ -626,7 +643,92 @@ filterButton.addEventListener('click', (event) => {
...
+      if (users && users.length > 0) {
+        users.forEach((user) => {
+          const userDiv = document.createElement('div');
+          userDiv.className = 'suggestion';
+          userDiv.textContent = user.username;
+
+          suggestionContainer.appendChild(userDiv);
+
+          userDiv.addEventListener('click', async () => {
+            userNameFilterInput.value = user.username;
```
Comment (Achintya-Chatterjee) [inline requests/script.js:693]: We should optimize debounce logic to prevent redundant API calls.
 - The current implementation clears the timeout but still allows multiple requests if the user types quickly.  
 - This can cause unnecessary load on the server and degrade performance.  
- Consider using _.debounce() from lodash or write a custom debounce function with a trailing execution strategy.  
- This ensures the API call is made only after the user stops typing, improving efficiency.  
- Additionally, caching previous results for the same input can further reduce API requests.

File: requests/script.js [hunk 4]
```diff
@@ -626,7 +643,92 @@ filterButton.addEventListener('click', (event) => {
...
+
+            const requestData = { requestedBy: user.username };
+            await renderRequestCards(requestData);
+          });
+        });
+      }
+    }, 300);
+  } else {
+    suggestionContainer.classList.add('hidden');
+    suggestionContainer.innerHTML = '';
```
Comment (Achintya-Chatterjee) [inline requests/script.js:705]: Clearing and repopulating `suggestionContainer (suggestionContainer.innerHTML = '';`) is an **anti-pattern**. Instead, We should use `DocumentFragment` to append elements in one go.
Also, if users re-type the same input, you're re-fetching data instead of caching previous results

File: requests/util.js [hunk 5]
```diff
@@ -43,6 +43,9 @@ function getQueryParamsString(requestType, query) {
   ) {
     queryParam += `&state=${query.state}`;
   }
+  if (query?.requestedBy != null) {
+    queryParam += `&requestedBy=${query.requestedBy}`;
+  }
```
Comment (Achintya-Chatterjee) [inline requests/util.js:48]: The query parameters are concatenated directly (`queryParam += '&requestedBy=' + query.requestedBy;`). If `queryParam` is empty, the query string starts with `&`instead of `?`. So we should use `new URLSearchParams(query).toString()` instead of manually appending query parameters.

File: requests/script.js [hunk 6]
```diff
@@ -540,30 +552,35 @@ async function performAcceptRejectAction(isAccepted, e) {
       state: isAccepted ? 'APPROVED' : 'REJECTED',
     });
   }
+
   const parentDiv = e.target.closest('.request__card');
-  parentDiv.classList.add('disabled');
-  const removeSpinner = addSpinner(parentDiv);
+  requestPageContainer.classList.add('disabled');
+  const removeSpinner = addSpinner(requestPageContainer);
```
Comment (Achintya-Chatterjee) [inline requests/script.js:558]: You are disabling the entire `requestPageContainer (requestPageContainer.classList.add('disabled')`) instead of only disabling the specific card. So this means that one action blocks all other interactions, which is a poor UX decision. Instead, we should only disable the clicked request card.

File: requests/util.js [hunk 7]
```diff
@@ -151,3 +154,26 @@ function addRadioButton(labelText, value, groupName) {
...
+      headers: {
+        'Content-type': 'application/json',
+      },
+    });
+    const data = await res.json();
+    return data.users;
+  } catch (error) {
+    console.log(error);
+  }
+}
```
Comment (Achintya-Chatterjee) [inline requests/util.js:179]: If the server returns a non-200 response, you are not checking `res.ok`, which may lead to unexpected behaviours.

File: requests/script.js [hunk 8]
```diff
@@ -105,9 +111,15 @@ function updateUrlWithQuery(type) {
...
-    (nextLink || '/requests' + getQueryParamsString(requestType, query));
+  let finalUrl;
+  if (query?.state?.[0] || query?.requestedBy) {
+    finalUrl =
+      API_BASE_URL + '/requests' + getQueryParamsString(requestType, query);
+  } else {
+    finalUrl =
+      API_BASE_URL +
+      (nextLink || '/requests' + getQueryParamsString(requestType, query));
+  }
```
Comment (Achintya-Chatterjee) [inline requests/script.js:122]: `getRequests()` function has redundant conditions. `if (query?.state?.[0] || query?.requestedBy) {...}` should be refactored. Why are we doing repetitive URL assignments, this is leading to unnecessary complexity.

File: requests/script.js [hunk 9]
```diff
@@ -626,7 +643,92 @@ filterButton.addEventListener('click', (event) => {
...
+  if (!selectedState) {
+    requestContainer.innerHTML = '';
+    await renderRequestCards(requestData);
+  } else {
+    requestContainer.innerHTML = '';
+    requestData.state = selectedState;
+    await renderRequestCards(requestData);
+  }
+});
```
Comment (Achintya-Chatterjee) [inline requests/script.js:667]: The function calls `renderRequestCards()` twice, once for `requestData.state === null` and once for `state !== null`. Instead of doing this, we should structure the logic cleanly.

**Review by Achintya-Chatterjee — CHANGES_REQUESTED** [review]

File: requests/util.js (hunk 7 above)
Comment (AnujChhikara) [reply requests/util.js]: added that check

File: requests/script.js (hunk 8 above)
Comment (AnujChhikara) [reply requests/script.js]: Refactored this function

File: requests/script.js (hunk 6 above)
Comment (AnujChhikara) [reply requests/script.js]: Improved this, now only that card will get disabled when we perform action on that request card

File: requests/script.js [hunk 10]
```diff
@@ -626,7 +641,94 @@ filterButton.addEventListener('click', (event) => {
...
+        suggestionContainer.appendChild(fragment);
+      }
+    }, 300);
+  } else {
+    suggestionContainer.classList.add('hidden');
+    while (suggestionContainer.firstChild) {
+      suggestionContainer.removeChild(suggestionContainer.firstChild);
+    }
+  }
+});
```
[the remark this answers is not in the document]
Comment (AnujChhikara) [reply requests/script.js:707]: The getUsersByUsername function already has error handling that logs issues and returns an empty array on failure, so the app won’t crash.

**Comment by AnujChhikara** [comment]
Hi @Achintya-Chatterjee, thanks for the review! I’ve addressed most of your comments. Regarding your suggestion to cache API calls, we currently don’t have a caching mechanism on the dashboard site. There are several options—like using a Map, a custom function, or other approaches—but I’d love your input on how to choose the best one for our use case. How do you think we should decide?

File: requests/script.js (hunk 3 above)
Comment (Achintya-Chatterjee) [reply requests/script.js]: we should write a custom debounce function, I think this will be the most sought-after approach.

Comment (AnujChhikara) [reply requests/script.js]: Okay, we already have a custom function for debounce , I have used that one.

Comment (Achintya-Chatterjee) [reply requests/script.js]: Cool

File: __tests__/requests/requests.test.js [hunk 11]
```diff
@@ -370,5 +381,51 @@ describe('Tests the request cards', () => {
...
+      for (const card of requestCards) {
+        const statusText = await card.$eval(
+          '[data-testid="request-status"]',
+          (el) => el.textContent,
+        );
+        expect(statusText).toContain('Approved');
+      }
+    });
+
+    it('should show user suggestions after debounced input', async () => {
```
Comment (Achintya-Chatterjee) [inline __tests__/requests/requests.test.js:412]: the test correctly verifies that suggestions appear and include the expected username, it would be beneficial to add assertions for edge cases such as empty results, special characters, or minimum character threshold.

**Review by Achintya-Chatterjee — CHANGES_REQUESTED** [review]

File: __tests__/requests/requests.test.js (hunk 11 above)
Comment (AnujChhikara) [reply __tests__/requests/requests.test.js:436]: added please check

**Review by Achintya-Chatterjee — APPROVED** [review]

**Review by iamitprakash — APPROVED** [review]
